pyomo/util/latex_printer.py

```diff
@@ -0,0 +1,1336 @@
+                'Formulation is infeasible due to bounds on variable %s' % (vr.name)
+            )
+        elif lowerBoundValue == 1:
+            lowerBound = ' = 1 '
+        elif lowerBoundValue > 0:
+            lowerBound = str(lowerBoundValue) + ' \\leq '
+        else:
+            lowerBound = ' 0 \\leq '
+
+        if upperBoundValue < 0:
+            raise InfeasibleConstraintException(
+                'Formulation is infeasible due to bounds on variable %s' % (vr.name)
+            )
+        elif upperBoundValue == 0:
+            upperBound = ' = 0 '
+        elif upperBoundValue < 1:
+            upperBound = ' \\leq ' + str(upperBoundValue)
+        else:
+            upperBound = ' \\leq 1 '
+
+    else:
+        raise DeveloperError(
+            'Invalid domain somehow encountered, contact the developers'
+        )
+
+    varBoundData = {
+        'variable': vr,
+        'lowerBound': lowerBound,
+        'upperBound': upperBound,
+        'domainName': domainName,
+        'domainLatex': domainMap[domainName],
+    }
+
+    return varBoundData
+
+
+def multiple_replace(pstr, rep_dict):
+    pattern = re.compile("|".join(rep_dict.keys()), flags=re.DOTALL)
+    return pattern.sub(lambda x: rep_dict[x.group(0)], pstr)
+
+
+def latex_printer(
+    pyomo_component,
+    latex_component_map=None,
+    write_object=None,
+    use_equation_environment=False,
+    explicit_set_summation=False,
+    throw_templatization_error=False,
+):
+    """This function produces a string that can be rendered as LaTeX
+
+    As described, this function produces a string that can be rendered as LaTeX
+
+    Parameters
+    ----------
+    pyomo_component: _BlockData or Model or Objective or Constraint or Expression
+        The Pyomo component to be printed
+
+    latex_component_map: pyomo.common.collections.component_map.ComponentMap
+        A map keyed by Pyomo component, values become the latex representation in
+        the printer
+
+    write_object: io.TextIOWrapper or io.StringIO or str
+        The object to print the latex string to.  Can be an open file object,
+        string I/O object, or a string for a filename to write to
+
+    use_equation_environment: bool
+        If False, the equation/aligned construction is used to create a single
+         LaTeX equation.  If True, then the align environment is used in LaTeX and
+         each constraint and objective will be given an individual equation number
+
+    explicit_set_summation: bool
+        If False, all sums will be done over 'index in set' or similar.  If True,
+        sums will be done over 'i=1' to 'N' or similar if the set is a continuous
+        set
+
+    throw_templatization_error: bool
+        Option to throw an error on templatization failure rather than
+        printing each constraint individually, useful for very large models
+
+
+    Returns
+    -------
+    str
+        A LaTeX string of the pyomo_component
```

Comment:
I am not sure I am in favor of this method always returning a big string. Maybe only return the big string if the user didn't provide an `ostream`?

This is a special case of a concern I have with the bulk of the method below where we are building up a single HUGE string.  Because Python strings are immutable, every time you `+=`, you are allocating a new string that is a copy of the old one with the extra characters.  This leads to "O(n^2)" behavior that can cause this to be very slow for large models.  I would rather this be structured to build up (and export) smaller chunks as it goes.